### ISSUE TITLE
HPC clusters: login-node env may not match compute-node architecture (GPU benchmarking + env setup)

### ISSUE THREAD

**Issue by sambit-giri** [cosmic-reionization/BEoRN#44]
## Problem: login-node environment ≠ compute-node environment on HPC clusters

Ran into this concretely while validating GPU/autodiff painting (issue #42): the `beorn` conda env on this cluster's **login node** is `x86_64` with a locally-attached NVIDIA L40 GPU — but that GPU is **shared across every logged-in user** (127 users on
this node), so any timing/benchmarking done there is dominated by contention, not the code. Meanwhile the cluster's actual SLURM-scheduled GPU partition is a completely different architecture (`aarch64`, NVIDIA GH200 "Grace-Hopper"), which the x86_64 env cannot run on at all (`Exec format error`).

This isn't specific to this cluster — it's a general HPC pattern worth documenting for anyone running BEoRN elsewhere:

- **Login nodes** are for editing/building/submitting jobs, often x86_64, sometimes with a small GPU for interactive dev — but shared by everyone logged in.
- **Compute/job nodes** (reached only via `sbatch`/`srun`) are the real resource,  may be a **different CPU architecture** (x86_64 vs aarch64 Grace, or even different  x86_64 microarchitectures with different available instruction sets), and are
  allocated exclusively (or fairly) per job.
- A conda/venv environment built on the login node is not guaranteed to run — or run  *well* — on a compute node, and vice versa. Any performance number measured on a login node should be treated as unreliable for anything involving contended shared hardware (GPU, sometimes even CPU cores).

## What to check before benchmarking or building an env

```bash
uname -m                       # login node arch
srun -p <gpu-partition> --gpus=1 --time=00:05:00 uname -m   # compute node arch
sinfo -o "%P %G"               # partitions + GPU types
scontrol show node <name> | grep -i arch   # per-node architecture
nvidia-smi --query-compute-apps=pid,used_memory --format=csv   # who else is on a shared GPU
```

If the two `uname -m` outputs differ, **the login-node env cannot be used on compute nodes** (or vice versa) — building a separate environment for the compute architecture is required.

## Recipe: building an architecture-matched environment for BEoRN runs

1. **Check what the site already provides.** HPC centers often ship pre-built, architecture-native bundles (`module avail`) that include the hard-to-build scientific stack (numpy/scipy/mpi4py compiled against the site's MPI/interconnect, sometimes pandas/Cython too). Reuse these rather than rebuilding from source — they're usually better-optimized and validated by the site than a generic pip wheel, and for `mpi4py` in particular you *want* the site's build (linked against the real interconnect), not a generic PyPI wheel that bundles its own MPICH.

2. **Create the new env on a compute node of the target architecture** — either via an interactive `srun` session or inside a short debug/build batch job — never just cross-install from the login node blind. A `python -m venv --system-site-packages <path>` layered on top of the site module inherits its packages while staying isolated and pip-installable.

3. **Dry-run wheel availability before committing to a full install.** `pip install --dry-run --no-deps <package>` resolves and downloads metadata (and the wheel itself, for sizing) without installing — cheap way to confirm e.g. `torch`, `jax[cuda12]` (and specifically `jaxlib` + `jax-cuda12-plugin` + `jax-cuda12-pjrt`, which is where GPU support actually lives) have wheels for the target architecture before spending a full job on it. If a dependency has no binary wheel for the architecture, pip will fall back to a source build — which can be slow, need extra system dev packages, or fail outright; better to catch that up front.

4. **Install BEoRN's own dependencies deliberately, not automatically.** If the site module already provides a package at a slightly older version than BEoRN's `pyproject.toml` pin (e.g. `mpi4py>=4.1.1` vs the site's `4.1.0`), letting a plain `pip install -e .` resolve dependencies can silently *upgrade* that package to a generic PyPI build and detach it from the site's MPI stack. Prefer:
   ```bash
   pip install <each-missing-package>    # only what's actually missing
   pip install --no-deps -e /path/to/BEoRN   # install BEoRN itself without touching deps
   ```
   and manually verify the site-provided versions satisfy BEoRN's constraints closely enough (a one-patch-version gap like 4.1.0 vs 4.1.1 is usually fine; anything larger deserves a closer look).

5. **Verify GPU visibility explicitly**, not just "the install succeeded":
   ```python
   import torch; print(torch.cuda.is_available(), torch.cuda.get_device_name(0))
   import jax; print(jax.devices())
   ```
   Both frameworks can install cleanly yet silently fall back to CPU if the CUDA plugin didn't match the driver/architecture.

6. **Watch out for environments that get moved after creation.** `venv`-created activation scripts (`bin/activate`) hardcode the **absolute path** of the venv at creation time — they do not resolve it dynamically. Moving/renaming the venv's parent directory breaks `source bin/activate` (it'll point PATH at a now-nonexistent location) even though the venv's own files are otherwise intact. Decide the final location *before* building, or budget for a quick rebuild (`rm -rf` + re-run the build script — pip's local wheel cache makes this fast the second time) after any move.

## Example scripts

Generic, ready-to-copy templates implementing the recipe above are provided in [`scripts/`](https://github.com/cosmic-reionization/BEoRN/tree/main/scripts) (see `scripts/README.md`, "HPC setup scripts" section):

| Script | Purpose |
|---|---|
| `hpc_setup_01_check_node_environment.sh` | Reports the compute node's CPU architecture, GPU visibility/contention, and what a site-provided Python module (if any) already includes. |
| `hpc_setup_02_check_wheel_availability.sh` | Dry-run `pip install` checks (no install) that BEoRN's GPU-capable dependencies — especially `jaxlib` / `jax-cuda12-plugin` / `jax-cuda12-pjrt`, the pieces that actually give jax GPU support — have prebuilt wheels for this architecture, before committing to a full build. |
| `hpc_setup_03_build_beorn_env.sh` | Builds the environment: a venv layered on the site's Python module (if any), the missing core deps, `torch`/`jax[cuda12]`, and an editable `--no-deps` install of BEoRN itself. Ends with a GPU-visibility check for both frameworks. |
| `hpc_setup_04_run_output_consistency.sh` | Validates the environment built in step 3 by running the MPI-parallel `tests/test_output_consistency.py` as a SLURM job. |

All four use the same placeholder convention (`YOUR_HPC_ACCOUNT`,
`/path/to/...`) — copy them out, fill in the placeholders for your cluster (account, partition, site module name if any, and the environment's final path — see the "watch out for environments that get moved" point above), and run in order. They don't need to live inside a BEoRN checkout to be used.

Once an environment exists, `scripts/` also has a second, "HPC job examples" section (`hpc_example_job_run.sh` / `hpc_example_job_notebook.sh` / `hpc_example_job_interactive.sh`) with templates for day-to-day use, including how to `sbatch` a job and keep working in a notebook interactively against it — see `scripts/README.md` for that section.

**Comment by sambit-giri** [comment]
## Arrhenius cluster (NAISS) case

Concrete walk-through of the recipe above, for reference.

**The mismatch.** The `beorn` conda env used interactively lives on the login node — `x86_64`, with a locally-attached NVIDIA L40 GPU. That GPU is **shared across every logged-in user** (127 users on the node at the time), so a `paint_single` timing comparison across `fft_backend='numpy'/'jax'/'torch'` gave numbers that didn't even self-consistently add up (e.g. one backend's own reported sub-step timings summed to a fraction of its reported total — classic contention symptom, not a code issue). The cluster's actual SLURM-scheduled GPU partition, meanwhile, is `aarch64` (NVIDIA GH200 "Grace-Hopper") — a different architecture the x86_64 env cannot run on at all (`Exec format error` on a direct interpreter invocation).

**Following the recipe:**

1. `sinfo -o "%P %G"` / `scontrol show node <n>` confirmed the `gpu` partition nodes report `Arch=aarch64`, vs. the login node's `x86_64` — architecture mismatch confirmed before touching anything else.
2. A short debug job (`--gpus=1`, ~15 min) on the `gpu` partition showed the node fully dedicated (0% GPU util, no other processes) and that the site's `GPU/Python/3.13.5-bundle-SciPy-2025.07-mpi4py-4.1.0-gcc-2025b-eb` module already provides native `aarch64` `numpy 2.3.2`, `scipy 1.16.1`, `mpi4py 4.1.0`, `pandas 2.3.1` — h5py/astropy/matplotlib/Cython were missing.
3. `pip install --dry-run --no-deps <pkg>` for `torch`, `jaxlib`, `jax-cuda12-plugin`, `jax-cuda12-pjrt`, and the missing core deps all resolved clean `manylinux_..._aarch64` wheels — no source builds needed anywhere, including for the CUDA-enabled pieces specifically.
4. Built the environment: `python -m venv --system-site-packages`, then `pip install h5py astropy matplotlib tqdm pyyaml torch "jax[cuda12]" tools21cm`, then `pip install --no-deps -e /path/to/BEoRN` (deliberately `--no-deps`, since BEoRN's `pyproject.toml` pins `mpi4py>=4.1.1` and the site module provides `4.1.0` — a plain dependency-resolving install would have tried to upgrade `mpi4py` to a generic PyPI build, detaching it from the site's real interconnect, for a one-patch-version gap that doesn't matter functionally here).
5. Verified explicitly rather than trusting "install succeeded": `torch.cuda.is_available() == True`, `torch.cuda.get_device_name(0) == 'NVIDIA GH200 120GB'`, `jax.devices() == [CudaDevice(id=0)]`, plus an actual matmul on each to rule out a plugin that loads but silently no-ops.
6. **Hit the "moved after creation" gotcha directly.** The environment was first built at a path *inside* the git checkout, then moved out to a sibling directory (correctly — build output shouldn't live inside the package tree). `bin/activate` kept pointing at the old, now-deleted path; `source`-ing it set `PATH` to a location that no longer existed, silently. Fix was the one described above: `rm -rf` the venv and rebuild fresh at the final location (a couple of minutes, since pip's wheel cache made the large downloads — `torch` 427 MB, `jax-cuda12-pjrt` 170 MB — instant the second time) rather than trying to patch the hardcoded paths.

**Result.** Re-running the `artificial_halo_gpu.ipynb` validation notebook on the dedicated GH200 node gave clean, internally-consistent numbers (the overlap-photon-spreading CPU step — the actual wall-time bottleneck in `paint_single`, unrelated to `fft_backend` — now measured flat at 3.42–3.48 s across all three backends, instead of varying 5×–8× depending on what else happened to be on the shared login-node GPU at the time), while reproducing the same gradient/physics numbers (autodiff vs. finite-difference checks, closed-form-ODE vs. `solve_ivp` agreement) measured earlier on the login node — confirming those were correct all along and only the *timing* story needed the dedicated hardware.

**Comment by sambit-giri** [comment]
Implemented in commit 9060a8c: the `scripts/` folder now has the generic `hpc_setup_01`-`04` environment setup/validation sequence and the `hpc_example_job_run`/`notebook`/`interactive` day-to-day job templates described above.
